src/sparseml/modifiers/obcq/utils/layer_compressor.py

```diff
@@ -0,0 +1,259 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import logging
+from typing import Dict, List
+
+import torch
+import torch.nn as nn
+from torch.nn import Module
+
+from sparseml.modifiers.obcq.utils.sparsegpt import SparseGPT
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class LayerCompressor:
+    """
+    Runs the SparseGPT algorithm on a single layer using calibration data inputs
+
+    Lifecycle:
+        - compress
+            - pre_compress
+            - fasterprune
+            - post_compress
+
+    :param model: model containing the layer we are running compression on
+    :param layer: layer to run compression on
+    :param layer_index: index of layer in the model
+    :param inputs: calibration data to pass through the layer
+    :param args: additional keyword arguments
+    """
+
+    def __init__(
+        self, model: Module, layer: Module, layer_index: int, inputs: List, args: Dict
+    ):
+        self.model = model
+        self.layer = layer
+        self.layer_index = layer_index
+        self.inputs = inputs
+        self.args = args
+
+    def compressible_modules(self) -> Dict:
+        """
+        Get the list of modules in the layer that can be compressed
+
+        :return: dictionary of compressible modules
+        """
+        quantize = self.args.get("quantize", False)
+        if quantize:
+            # The layer names are changed due to quantization modifiers, therefore
+            # we need a slightly different func to retrieve layers
+            modules = _find_quant_layers(self.layer)
+        else:
+            modules = _find_layers(self.layer)
+        return modules
+
+    def pre_compress(self, **kwargs) -> Dict:
+        """
+        Sets up the SparseGPT objects for each compressible module, computes the Hessian
+        for each using the calibration data.
+
+        :return: gpt objects for each module if not updating sequentially, otherwise
+        a blank dictionary
+        """
+        if not self.args["sequential_update"]:
+            subset = self.compressible_modules()
+
+            gpts = {}
+            for name in subset:
+                gpts[name] = SparseGPT(subset[name])
+
+            def add_batch(name):
+                def tmp(_, inp, out):
+                    gpts[name].add_batch(inp[0].data, out.data)
+
+                return tmp
+
+            handles = []
+            for name in gpts:
+                handles.append(subset[name].register_forward_hook(add_batch(name)))
```

Comment:
will this work with FSDP? Not sure which hooks are supported - could we do this with callbacks instead?